client/session.go

```diff
@@ -0,0 +1,277 @@
+	totalBatches := topologyMap.Replicas() *
+		int(math.Ceil(float64(s.opts.GetWriteOpPoolSize())/float64(s.opts.GetWriteBatchSize())))
+	hostBatches := int(math.Ceil(float64(totalBatches) / float64(topologyMap.HostsLen())))
+	writeBatchRequestPool := newWriteBatchRequestPool(hostBatches)
+	writeRequestArrayPool := newWriteRequestArrayPool(hostBatches, s.opts.GetWriteBatchSize())
+	return newHostQueue(host, writeBatchRequestPool, writeRequestArrayPool, s.opts)
+}
+
+func (s *session) setWriteOpPoolsWithLock(topologyMap m3db.TopologyMap) {
+	totalShards := len(topologyMap.ShardScheme().All().Shards())
+	if len(s.writeOpPools) == totalShards {
+		// Already created all pools
+		return
+	}
+
+	s.writeOpPools = make([]writeOpPool, totalShards)
+	totalSize := s.opts.GetWriteOpPoolSize()
+	eachSize := totalSize / totalShards
+	for i := range s.writeOpPools {
+		s.writeOpPools[i] = newWriteOpPool(eachSize)
+	}
+}
+
+func (s *session) Write(id string, t time.Time, value float64, unit xtime.Unit, annotation []byte) error {
+	var (
+		w             *writeOp
+		pool          writeOpPool
+		wg            sync.WaitGroup
+		enqueueErr    error
+		enqueued      int
+		resultErrLock sync.Mutex
+		resultErr     error
+		resultErrs    int
+		quorum        int
+	)
+
+	timeType, timeTypeErr := convert.UnitToTimeType(unit)
+	if timeTypeErr != nil {
+		return timeTypeErr
+	}
+
+	ts, tsErr := convert.TimeToValue(t, timeType)
+	if tsErr != nil {
+		return tsErr
+	}
+
+	s.RLock()
+	routeErr := s.topoMap.RouteForEach(id, func(shard uint32) {
+		pool = s.writeOpPools[shard]
+		w = pool.Get()
+		w.request.ID = id
+		w.datapoint.Value = value
+		w.datapoint.Timestamp = ts
+		w.datapoint.TimestampType = timeType
+		w.datapoint.Annotation = annotation
+		w.completionFn = func(result interface{}, err error) {
+			if err != nil {
+				resultErrLock.Lock()
+				if resultErr == nil {
+					resultErr = err
+				}
+				resultErrs++
+				resultErrLock.Unlock()
+			}
+			wg.Done()
+		}
+	}, func(idx int, host m3db.Host) {
+		wg.Add(1)
+		enqueued++
+		if err := s.queues[idx].Enqueue(w); err != nil && enqueueErr != nil {
+			// NB(r): if this ever happens we have a code bug, once we are in the read lock
+			// the current queues we are using should never be closed
+			enqueueErr = err
+		}
+	})
+	quorum = s.topoMap.QuorumReplicas()
+	s.RUnlock()
+
+	if routeErr != nil {
+		return routeErr
+	}
+
+	if enqueueErr != nil {
+		s.opts.GetLogger().Errorf("failed to enqueue write: %v", enqueueErr)
+		return enqueueErr
+	}
+
+	// Wait for writes to complete
+	wg.Wait()
+
+	// Return write to pool
+	pool.Put(w)
+
+	if resultErrs > 0 {
+		// Check consistency level satisfied
+		level := s.opts.GetConsistencyLevel()
+		success := enqueued - resultErrs
+		switch level {
+		case m3db.ConsistencyLevelAll:
+			return fmt.Errorf("failed to meet %s with %d/%d success", level.String(), success, enqueued)
+		case m3db.ConsistencyLevelQuorum:
+			if success >= quorum {
+				// Meets quorum
+				break
```

Comment:
Not sure if we should return an error here, semantically this could be considered as success per the consistency requirement, will it make more sense to just log the error but return nil to the client?